packages/repository/test/acceptance/relations/hasMany/models/customer.model.ts

```diff
@@ -0,0 +1,36 @@
+// Copyright IBM Corp. 2017,2018. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Entity, Class} from '../../../../../';
+import {Order} from './order.model';
+import {OrderRepository} from '../repositories/order.repository';
+import {Repository} from '../../../../../src/repositories';
+import {model, property, hasMany} from '../../../../../src/decorators';
+
+@model()
+export class Customer extends Entity {
+  @property({
+    type: 'number',
+    id: true,
+  })
+  id?: number;
+
+  @property({
+    type: 'string',
+  })
+  name: string;
+
+  @property({
+    type: 'string',
+    regexp: '^w+([.-]?w+)*@w+([.-]?w+)*(.w{2,3})+$',
+  })
+  email: string;
+
+  @hasMany({
+    target: 'Order',
+    as: 'orders',
+  })
+  customerOrders: OrderRepository;
```

Comment:
I am not sure if it's  good idea to have the same API for a relation repository as we have for regular models.

For example, with HasAndBelongsToMany relation, the repository needs to expose additional API like add/remove (link/unlink), see http://loopback.io/doc/en/lb3/HasAndBelongsToMany-relations.html

I think we will eventually need to have different repository interfaces/APIs:

 - `CrudRepository<Model, IdType>` interface and `DefaultCrudRepository<Model, IdType>` class, providing direct access to all instances of a model
 - `HasAndBelongsToManyRepository<TargetModel, ...>` interface and `DefaultHasAndBelongsToManyRepository<TargetModel, ...>` class, providing API to work with HasAndBelongsToMany relation
 - and similarly for other relation types

Reply:
I like that approach, but was thinking that's heading down the road of how we did it in LB3? 
```
For example, Customer and Order has the belongsTo relation. When customerId was given, it becomes the filter for the Order repository. It implies that Order repository is not a free form repository but is constrained by customerID=xxx.

When doing CRUD operation, make sure this constraint is honoured.
Currently, we are using primary key/foreign key for the relationship. In LB Next, we should allow people to use any properties.
```
My thinking was that we wanted a new type of relation engine where, based on the relation metadata and constraints, we build a constrained version of the target repository using a factory method. The method would also be responsible for scoping the CRUD methods allowed on the target repository.
cc @raymondfeng

Reply:
I think my comment/proposal is compatible with the proposal you quoted in your comment.

In my proposal, I am mostly concerned about the repository API we are going to expose for relations. As the `BelongsTo` relation shows, this API cannot be the same one as we have in `CrudRepository`.

As I see the part you are quoting, it's concerned about the implementation for that repository API - how are we going to implement methods like `order.customer.create`, allowing people to specify any properties to constrain the link between the related models.

While we are touching this point, there is another use case in LB3 that's similar to relations, although it deals with a single model only: scopes. See https://loopback.io/doc/en/lb3/Model-definition-JSON-file.html#scopes

Reply:
Thank you for clarifying @bajtos. That makes sense! I'll check out scopes as well.